### DISCUSSION TITLE
Only show if xamarin projects

### DISCUSSION THREAD

**Discussion by ekrogh** [VsixCommunity/Community.VisualStudio.Toolkit#147] [Q&A]
Hi

I’m creating an extension that (so far) only works on xamarin forms solutions. It is activated via the Solution Explorer Context Menu for the solution.

Is it possible to only show the menu button when the solution is xamarin forms ? E.i. are the any visibility constraints I can set on the menu button for my extension that allow it to be only shown if it is a xamarin forms solution ?

—
Eigil

**Comment by CZEMacLeod** [comment]
Not sure specifically about detecting a Xamarin Forms solution (I assume you mean a solution that contains a Xamarin Forms project, as it could contain many other business logic, data access and other projects, as well as the native platform 'head' projects).
But I think you could override the `BeforeQueryStatus` method for the command, and adjust the visibility of the menu item. It isn't the same, but there is a discussion of this in the [Writing Visual Studio Extensions with Mads - Dynamically add menu commands](https://youtu.be/mylKmRKF2Rs) YouTube video.
You would need to scan the open solution and do your detection. You could probably do this using solution/project load events and cache the state which would then be used.
Remember that with MAUI, things will be changed up quite a bit, so your detection would have to be a bit different.

**Comment by mrlacey** [comment]
I have some code that identifies if a project is WPF, UWP, WinUI, Xamarin.Forms, or MAUI at https://github.com/mrlacey/Rapid-XAML-Toolkit/blob/main/VSIX/RapidXaml.Shared/VisualStudioIntegration/VisualStudioAbstraction.cs#L47

It's not the prettiest and will need some modification for your needs but may be a starting point for you

**Comment by ekrogh** [comment]
Thank you for the reply!

 

Maybe I'm confused about Solution / Project.

 

When you start a xamarin forms Solution / Project you go into

File -> New -> Project

and select

Mobile App (Xamarin.Forms)

 

And then a Solution is created with up to 3 Projects (Android, iOS and UWP) and more projects can be added later.

 

I thought that it might later be possible to detect that this is the kind of Solution / Project.

 

--

Eigil

**Reply by mrlacey** [reply]
In theory, a solution can have any number of projects in it and a project can be of different types.
To further complicate things, a project can be of multiple types.
For example, if I create a Xamarin.Forms project that contains Android, iOS, and a UWP head, is the UWP project a UWP project or a Xamarin.Forms project? Or both?
Then consider I add a WebAPI project to the same solution. The solution can no longer be considered **only** a Xamarin.Forms solution.

For my purpose, I mostly ignore solutions that have multiple types of projects (i.e. Xamarin.Forms apps and an associated WebAPI backend) in the same solution and do a best effort identify the "type" of project that's open by looking at all the projects in the solution.

**Comment by madskristensen** [comment]
You can do this with a custom UI rule and VisibilityConstraints. I've got videos on it on the YouTube Playlist. Sorry, out on leave and not by PC

**Reply by mrlacey** [reply]
This one: https://youtu.be/Pf6byQ1xQQc

**Comment by ekrogh** [comment]
FYI: Here is my solution:

The Package class is decorated like this:
```
	[ProvideUIContextRule
		(
			PackageGuids.uiContextSupportedFlavorsString,
			name: "Auto show",
			expression: "(single | multiple) & (UWP | XamarinAndroid | XamariniOS | XamarinmacOS)",
			termNames: new[]
				{
					"single"
					,
					"multiple"
					,
					"UWP"
					,
					"XamarinAndroid"
					,
					"XamariniOS"
					,
					"XamarinmacOS"
				},
			termValues: new[]
				{
					UICONTEXT.SolutionHasSingleProject_string
					,
					UICONTEXT.SolutionHasMultipleProjects_string
					,
					"SolutionHasProjectFlavor:A5A43C5B-DE2A-4C0C-9213-0A381AF9435A"
					,
					"SolutionHasProjectFlavor:EFBA0AD7-5A72-4C68-AF49-83D382785DCF"
					,
					"SolutionHasProjectFlavor:6BC8ED88-2882-458C-8E55-DFD12B67127B"
					,
					"SolutionHasProjectFlavor:A3F8F2AB-B479-4A4A-A458-A89E7DC349F1"
				}
		)
	]

```

**Reply by CZEMacLeod** [reply]
This only detects the head projects - so it will show up for Android Native, UWP and other projects, regardless of whether Xamarin.Forms is involved or not. It also misses out some of the other available Xamarin forms head project types. (Tizen and WPF at least).
It also is not compatible with MAUI which is going to replace Xamarin.Forms soon.

If this works for your use case then great, however I don't think it really addresses the original question, and perhaps isn't/shouldn't be the 'answer' from a toolkit perspective.

@mrlacey's solution to detect the installed nuget packages and detect if `Xamarin.Forms` is installed is a much better bet, I'm not sure if this can be used in any way with `ProvideUIContextRule` but I would be much more inclined to go with that as a solution, and it actually detects the scenario requested, and can be updated in the future to detect the MAUI packages.

I could see this also being used to detect legacy ASP.NET 4 and ASP.NET Core projects easily, and would probably be a useful extension to the toolkit in general, perhaps along with an easy way of using these rules to enable commands.

**Reply by mrlacey** [reply]
My code is based on identifying the project type once the extension has already loaded in a project that contains .xaml files. I couldn't find a way to filter the loading with a UIContextRule that would include projects of type WPF (Framework & CORE), UWP, Xamarin.Forms, WinUI3, & MAUI.

**Reply by ekrogh** [reply]
@CZEMacLeod 
I will add the GUIDS for Tizen , GTK, etc. later.
- and doesn't xamarin.forms not always have at least 2 projects ? so "single" should possibly be removed - though my extension would work fine for "native" XamarinAndroid apps (you can't have Android app w.o xamarin ?)

**Reply by CZEMacLeod** [reply]
@ekrogh You specifically said Xamarin.Forms - a Xamarin Android (Native) app does not have Xamarin.Forms. That is the difference. You can have an (optional) shared business logic project, and a single or multiple head projects, with all the UI being native, using 'Xamarin' on Android and iOS, and then just normal UWP or similar for windows.
If you add Xamarin.Forms you then add a common UI project and use a layer to render the Xamarin.Forms flavour of XAML and its controls on each platform.
If you want to detect and run based on the head project types, that's fine, but I wanted to make sure that anyone else searching for a similar solution, didn't make an assumption that wasn't stated and go in the wrong direction.
As I said before, this will make your commands visible for a solution that has just a standard UWP windows app, nothing to do with Xamarin or Xamarin Forms. 
Also with .Net6, the Xamarin native projects just become .net6 target flavours, and indeed, you could end up with a single project with all the logic, and the native UI implementations for each platform in it. If MAUI is added (the equivalent of Xamarin.Forms) then your single project can have the shared UI, shared logic and all the targets.
